### PR TITLE
Fix Server WMS unstable feature IDs

### LINKED ISSUE #41124: Server WMS GetFeatureInfo feature ID inconsitencies
QGIS Server GetFeatureInfo returns different feature IDs for the same request on different instances.

Scenario:
- provider postgres, standard ordinary spatial layer with int64 pk, indexes and all
- multiple server instances behind a load balancer

What happens here is that depending on which particular request hits the instances first, the feature IDs assigned by the data provider (which are incremental starting from 1) change. The FID->PK mapping is cached so **this issue won't be present when using a single instance**.

So, the same client issuing the same request might hit a different instance of the server, getting back results with different feature IDs for the same feature.

I know that it is possible to instruct some load balances to always route a certain IP to the same instance instead of using round-robin or other schemes, but I believe that we should have way to consistently assign the same feature ID in a `GetFeatureInfo` response.

This is a pretty serious bug that prevent to use QGIS server in production under heavy load.

I cannot see a simple solution here, we cannot change the internal representation of `QgsFeature::id()` (qlonglong) without rewriting QGIS from the ground up, what we can probably do is to implement a `QString QgsFeature::providerId()` that delegates to the data provider to produce a consistent string representation of the feature ID to be used in these situations.

The data provider will calculate the stable feature ID from the PKs in a way that it make it easy to retrieve the feature from its `providerId`.

Any word of wisdom or other idea?

CC: @rldhont @pblottiere @dmarteau @mhugent

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #41124

Fix for JSON and GML only: the other formats were already ok.

### CODE REVIEW

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 1]
```diff
@@ -2432,10 +2436,11 @@ namespace QgsWms
   {
     //qgs:%TYPENAME%
     QDomElement typeNameElement = doc.createElement( "qgs:" + typeName /*qgs:%TYPENAME%*/ );
-    typeNameElement.setAttribute( QStringLiteral( "fid" ), typeName + "." + QString::number( feat->id() ) );
+    const QString fid( QgsServerFeatureId::getServerFid( *feat, layer->dataProvider()->pkAttributeIndexes() ) );
+    typeNameElement.setAttribute( QStringLiteral( "fid" ), QStringLiteral( "%1.%2" ).arg( typeName, fid ) );
 
     QgsCoordinateTransform transform;
-    if ( layer && layer->crs() != crs )
+    if ( layer && layer->crs() != crs ) // cppcheck-suppress nullPointerRedundantCheck
```
Comment (m-kuhn) [inline src/server/services/wms/qgswmsrenderer.cpp:2443]: I think the cppcheck is correct:

- either we are sure there is a layer and we can remove the nullptr check here
- or we need to check for a nullptr on line 2439 as well

Comment (elpaso) [reply src/server/services/wms/qgswmsrenderer.cpp:2443]: ok

Comment (elpaso) [reply src/server/services/wms/qgswmsrenderer.cpp:2443]: Removing the check because it's already checked in client code

Comment (m-kuhn) [reply src/server/services/wms/qgswmsrenderer.cpp:2443]: Can you add a `Q_ASSERT` to make the assumption explicit?

Comment (elpaso) [reply src/server/services/wms/qgswmsrenderer.cpp:2443]: nope, the check is correct, there is a case when a nullptr layer is passed.

Comment (elpaso) [reply src/server/services/wms/qgswmsrenderer.cpp]: I've added the missing check for the newly introduced fid from data provider, we should be safe now.

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 2]
```diff
@@ -2460,7 +2470,7 @@ namespace QgsWms
       {
         try
         {
-          QgsRectangle transformedBox = transform.transformBoundingBox( box );
+          const QgsRectangle transformedBox = transform.transformBoundingBox( box );
           box = transformedBox;
```
Comment (m-kuhn) [inline src/server/services/wms/qgswmsrenderer.cpp:2474]:
```suggestion
          box = transform.transformBoundingBox( box );
```

Not sure it's worth the extra line :stuck_out_tongue_closed_eyes:

**Review by m-kuhn — APPROVED** [review]

**Comment by elpaso** [comment]
unrelated test failure

**Pull request merged**